aptos-move/mvhashmap/src/types.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_aggregator::delta_change_set::DeltaOp;
+use aptos_crypto::hash::HashValue;
+use std::sync::Arc;
+
+pub type TxnIndex = u32;
+pub type Incarnation = u32;
+pub type Version = (TxnIndex, Incarnation);
+
+/// Returned as Err(..) when failed to read from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVDataError {
+    /// No prior entry is found.
+    NotFound,
+    /// Read resulted in an unresolved delta value.
+    Unresolved(DeltaOp),
+    /// A dependency on other transaction has been found during the read.
+    Dependency(TxnIndex),
+    /// Delta application failed, txn execution should fail.
+    DeltaApplicationFailure,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeError {
+    /// No prior entry is found.
+    NotFound,
+    /// A dependency on other transaction has been found during the read.
+    Dependency(TxnIndex),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVDataOutput<V> {
+    /// Result of resolved delta op, always u128. Unlike with `Version`, we return
+    /// actual data because u128 is cheap to copy and validation can be done correctly
+    /// on values as well (ABA is not a problem).
+    Resolved(u128),
+    /// Information from the last versioned-write. Note that the version is returned
+    /// and not the data to avoid copying big values around.
+    Versioned(Version, Arc<V>),
+}
+
+/// Returned as Ok(..) when read successfully from the multi-version data-structure.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MVCodeOutput<M, X> {
+    /// Arc to the executable corresponding to the latest module, and an indicator
+    /// whether the executable corresponds to a storage-version module.
+    Executable((Arc<X>, bool)),
+    /// Arc to the latest module, together with its (cryptographic) hash. Note that
+    /// this can't be a storage-level module, as it's from multi-versioned code map.
+    Module((Arc<M>, Option<HashValue>)),
```

Comment:
Also I don't quite understand why are we distinguishing Executable and Module? Is the idea that one is for CompiledModule and the other is for a more resolved form? If so I feel like this is a implementation detail that needs to be filled out in move side in your future refactor and not sth this mvhashmap should care about?

Reply:
CompiledModule / blob are the same thing at this point - we will and should optimize soon to save serialization etc overhead, but afaik that's the only difference.
Executable corresponds to more resolved form, something that will live in a sort of "stable" memory, be linkable, executable, so in the most simple (because our representations currently are pretty basic) I think corresponds to the structs and functions that are currently indexed and linked that way. The first version of the executable can just hold those Arc's directly, I'm thinking - later we will optimize memory consumption & re-use, representation, etc.

Why does MVHashMap need to know - it only knows via the trait, and the trait now mainly exposes the size() -> this is so we can implement a basic memory accounting model to RLU or flush or something. Storage-version executables will be re-used across the blocks, things that got overwritten will be pruned out, and at some point we will also have eviction based on metadata we can track and sizes.

Currently, the simplest thing is to make VersionedCode hold these Executables, and manage their correspondence with the blob (CompiledModules) - it's actually very convenient, because Move-VM will have to generate the executables, but due to concurrency, if you read a blob/CompiledModule, you should not have a dependency on it or need to validate, only once you obtain an executable, because that's the code you'll end up running. There will be "store_executable" function in the trait so when Move-VM doesn't get the executable (because it doesn't exist), it will create an executable, and call store. Then it will read again, until it gets an executable and can proceed.

Then there is a detail of init_module but that works out because data_store does the staging there - those executables won't be stored in MVHashMap via trait, which avoids all the today's problems - these will be just within the context of the transaction and if txn is successful, then the output is propagated, and if we want as an optimization we can also pass the executables there. Anyway, this last part is minor detail.